src/materials/LinearSpring.C

```diff
@@ -0,0 +1,247 @@
+  x_orientation /= x_orientation.norm(); // Normalizing with length to get orientation
+
+  // Get y orientation of the spring in global coordinate system
+  RealGradient y_orientation = getParam<RealGradient>("y_orientation");
+  // Real dot = x_orientation(0) * y_orientation(0) + x_orientation(1) * y_orientation(1) +
+  //            x_orientation(2) * y_orientation(2);
+  Real dot = x_orientation * y_orientation;
+
+  // Check if x and y orientations are perpendicular
+  if (abs(dot) > 1e-4)
+    mooseError("Error in LinearSpring: y_orientation should be perpendicular to "
+               "the axis of the beam.");
+
+  // Calculate z orientation in the global coordinate system as a cross product of the x and y
+  // orientations
+  RealGradient z_orientation = x_orientation.cross(y_orientation);
+  // z_orientation(0) = (x_orientation(1) * y_orientation(2) - x_orientation(2) * y_orientation(1));
+  // z_orientation(1) = (x_orientation(2) * y_orientation(0) - x_orientation(0) * y_orientation(2));
+  // z_orientation(2) = (x_orientation(0) * y_orientation(1) - x_orientation(1) * y_orientation(0));
+
+  // Calculate the rotation matrix from global to spring local configuration at t = 0
+  _original_global_to_local_rotation(0, 0) = x_orientation(0);
+  _original_global_to_local_rotation(0, 1) = x_orientation(1);
+  _original_global_to_local_rotation(1, 0) = y_orientation(0);
+  _original_global_to_local_rotation(1, 1) = y_orientation(1);
+  _original_global_to_local_rotation(0, 2) = x_orientation(2);
+  _original_global_to_local_rotation(1, 2) = y_orientation(2);
+  _original_global_to_local_rotation(2, 0) = z_orientation(0);
+  _original_global_to_local_rotation(2, 1) = z_orientation(1);
+  _original_global_to_local_rotation(2, 2) = z_orientation(2);
+
+  // _qp = 0
+  computeDeformations();
+
+  // computing spring forces
+  computeForces();
+
+  // computing spring stiffness matrix
+  computeStiffnessMatrix();
+}
+
+void
+LinearSpring::computeTotalRotation()
+{
+  _qp = 0;
+
+  // Currently this forumation is limited to small deformations in the spring,
+  // namely, it is assumed that there are no rigid body rotations in the spring,
+  // and that the total rotation matrix from global to local coordinates
+  // (calculated below) remains the same as the one at t = 0 throughout the
+  // duration of the simulation.
+  _total_global_to_local_rotation[_qp] = _original_global_to_local_rotation;
+}
+
+void
+LinearSpring::computeDeformations()
+{
+  // fetch the two end nodes for _current_elem
+  std::vector<Node *> node;
+  for (unsigned int i = 0; i < 2; ++i)
+    node.push_back(_current_elem->get_node(i));
+
+  // Fetch the solution for the two end nodes at time t
+  NonlinearSystemBase & nonlinear_sys = _fe_problem.getNonlinearSystemBase();
+  const NumericVector<Number> & sol = *nonlinear_sys.currentSolution();
+
+  // Calculating global displacements and rotations at the end nodes
+  RealVectorValue global_disp0(3, 0.0); // node 0
```

Comment:
RealVectorValue automatically has size 3 and initializes to zero. So Ben had suggested removing the (3, 0.0) in the beam code.

Reply:
Also, it might be better to create these vectors in the .h file and use it here than creating it at each iteration.